store.rb

```diff
@@ -238,23 +364,45 @@ def check_cache_age
   end
 
   def reload_instances()
-    # Here we use local time to preven race condition
-    # Basically cache fetch thread or zookeeper watch callback
-    # both will call this to refresh cache. Depending on which
-    # finishes first our cache will get set by the slower one.
-    # So in order to prevent setting cache to an older result,
-    # we set both cache and the timestamp of that version fetched
-    # Since timestamp will be monotonically increasing, we are
-    # sure that cache set will always have newer versions
-
-    fetch_start_time = Time.now
-    instances = load_instances_from_zk.freeze
-    @cache_mutex.synchronize do
-      if fetch_start_time > @cache_results_last_fetched_time then
-        @cache_results_last_fetched_time = fetch_start_time
-        @cache_results = instances
+
+    return unless @cache_mutex.try_lock
+
+    now = Time.now.to_i
+
+    if now > @cache_results_last_fetched_time + @cache_fetch_interval
+      inst, idx = load_instances
+
+      @cache_results = inst.freeze
+      @cache_indices = idx.freeze
+
+      case @opts['split_mode']
+      when 'store'
+        new_store = {
+          'inst' => @cache_results,
+          'idx'  => @cache_indices,
+        }
+        @cache_results_serialized = Oj.dump new_store
+      when 'server'
+        new_store = {
+          'examined' => 0,
+          'returned' => @cache_results.length,
+          'nodes'    => @cache_results,
+        }
+        @cache_results_serialized = Oj.dump new_store
       end
-    end
+
+
+      @cache_results_last_fetched_time = now
+      update_cache_fetch_interval
+
+      @log.info "reloaded cache. new reload interval = #{@cache_fetch_interval}"
+     end
+
+    @cache_mutex.unlock
```

Comment:
If there's a failure in reload_instances, will this mutex ever unlock? I'm thinking we might need to add a try ... finally here?